### ISSUE TITLE
optional argument matcher

### ISSUE THREAD

**Issue by VladaHejda** [mockery/mockery#301]
I want to suggest something like **OPTIONAL Matcher**, that will accept calling mocked method both with the optional argument and without it.

My code calls some method, sometimes with an optional argument, sometimes not:

``` php
$mock->method( $mandatoryArg, $optionalArg );
$mock->method( $mandatoryArg );
```

But for now, there is no way how to mock this by one specification, I have to do this:

``` php
$mock->shouldReceive( 'method' )
    ->with( 'mandatory', 'optional' );
$mock->shouldReceive( 'method' )
    ->with( 'mandatory' );
```

My suggestion is to mock it this way:

``` php
$mock->shouldReceive('method')
    ->with( 'mandatory', \Mockery::optional( 'optional' ) );
```

When optional parameter should be matched, this can work:

``` php
$mock->shouldReceive('method')
    ->with( 'mandatory', \Mockery::optional( \Mockery::any() ) );
```

_I found [similar issue](https://github.com/padraic/mockery/issues/167), but it requires slightly different behavior._

**Comment by padraic** [comment]
Closing issue as for matching. Should use Hamcrest and PR there if missing a specific need.

**Comment by aik099** [comment]
Unfortunately this is no ordinary matcher, because when mocking methods with optional arguments Mockery knows their default value (from reflection) and for that reason we can assume every `->with('madatoryValue`)` call to be equivalent to `->with('madatoryValue', `optionalValue`)`. Without this users have to pass `m::any()` for every optional parameter.

**Comment by padraic** [comment]
@aik099 Point taken. Reopening for now.

**Comment by robertbasic** [comment]
@padraic @davedevelopment is this still something we'd consider adding?

**Comment by tflori** [comment]
@aik099 I undertsand correct that `->shouldReceive('call')->with(m::type('string'), m::any())` matches `->call('something')` and `->call('anything', 'else')`?

**Comment by davedevelopment** [comment]
@robertbasic I'm on the fence, it's a very narrow use case.

**Comment by aik099** [comment]
`->shouldReceive('call')->with(m::type('string'), m::etcetra())` should match both of your calls.

**Comment by tflori** [comment]
@aik099 I don't know etcetera? where to find it?

**Comment by aik099** [comment]
Ah, I've mistakenly assumed this was a Mockery feature, but in fact it was Prophecy feature. I has `Argument::any` which is same as `m::any` but also has `Argument::cetera` which does match any number of arguments (including optional) after it's specified (see https://github.com/phpspec/prophecy/blob/d883566b83ae601a50b38b249e92450dc57cf875/src/Prophecy/Argument.php#L87). Maybe there is analog in Mockery, I have no idea.

**Comment by robertbasic** [comment]
@aik099 @tflori there is none, this whole ticket is about should we add it in or not.

**Comment by tflori** [comment]
then my opinion is for sure yes. I'm writing a library and offer a trait for easier mocking the library. I don't know if it is getting called with the second parameter and therefore I either have to omit the exact count (no way) or add for every method with optional parameters an `m::on(closure)` parameter. A optional parameter definition makes more sense in my opinion.

**Comment by tflori** [comment]
I just found out that it is not matching at all. You can not add an optional argument. The closure for `m::on(closure)` is not even called if the argument is not given. I thought the same like the guy from issue #752 that the closuer is called with all parameters.

Now for me there is still one question: You think about implementing an optional argument or not. That sounds like there is another option to realize this. You have a solution for optional arguments? How?

**Comment by robertbasic** [comment]
@tflori what do you mean the closure is not called if it's not given? The `m::on` is used to validate a single argument, not all of the arguments: http://docs.mockery.io/en/latest/reference/argument_validation.html#complex-argument-validation

**Comment by tflori** [comment]
if you omit the parameter the on closure is not called - how to allow different parameter sets for expectations?

**Comment by robertbasic** [comment]
This is now implemented with #860, closing.

**Comment by tflori** [comment]
Sorry, to digging this out but it is still one of the first links when searching for optional parameters in mockery...

I don't even remember how I solved that in the other library. But I stumbled again over this issue. Thanks to the `AndAnyOtherArgs` matcher I have a workaround now but I don't think that [this](https://github.com/mockery/mockery/pull/860/files#diff-6001b75d3eba48471dcbccfea0346ccdf95a312b473a6c186c9128b7bb1993f5R344) is a viable solution.

My workaround at the moment is an `Optional` matcher:
```php
use Mockery\Matcher\AndAnyOtherArgs;

class Optional extends AndAnyOtherArgs
{

    public function match(&$actual)
    {
        return is_null($actual) || $this->_expected->match($actual);
    }

    public function __toString()
    {
        return '[' . $this->_expected . ']';
    }
}
```

That is not really correct but as long as the tests are green :grin: 

IMO argument matching should be different:
```php
    public function matchArgs(array $args)
    {
        if ($this->isArgumentListMatcher()) {
            return $this->_matchArg($this->_expectedArgs[0], $args);
        }
        foreach ($this->_expectedArgs as $i => $expected) {
            if (!array_key_exists($i, $args)) {
                // as all further parameters have to be optional too we can return true if the first matcher where we
                // don't get an argument is an optional matcher
                return $expected instanceof OptionalArg;
            }
            if (!$this->_matchArg($expected, $args[$i])) {
                return false;
            }
        }
        
        // if we have more args the last matcher has to be multiple argument and every other argument has to match
        if (count($args) > count((array) $this->_expectedArgs)) {
            $expected = end($this->_expectedArgs);
            reset($this->_expectedArgs);
            if (!$expected || !$expected instanceof MultiArg) {
                return false;
            }

            foreach (slice($args, $i+1) as $arg) {
                if ($this->_matchArg($expected, $arg)) {
                    return false;
                }
            }
        }
        return true;
    }
```

With these new matchers:
```php
class OptionalArg extends MatcherAbstract
{
    public function match(&$actual)
    {
        return is_null($actual) || $this->_expected->match($actual);
    }

    public function __toString()
    {
        return '[' . $this->_expected . ']';
    }
}

class MultiArg extends OptionalArg
{
    public function match(&$actual)
    {
        return $this->_expected->match($actual);
    }

    public function __toString()
    {
        return '[...' . $this->_expected . ']';
    }
}
```
